cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -689,16 +697,26 @@ func (s *service) getContainerPids(ctx context.Context, container *runc.Containe
 	return pids, nil
 }
 
-func (s *service) forward(ctx context.Context, publisher shim.Publisher) {
+func (s *service) Events(ctx context.Context, _ *emptypb.Empty, streamer taskAPI.TTRPCTask_EventsServer) error {
 	ns, _ := namespaces.Namespace(ctx)
 	ctx = namespaces.WithNamespace(context.Background(), ns)
 	for e := range s.events {
-		err := publisher.Publish(ctx, runtime.GetTopic(e), e)
+		evt, err := protobuf.MarshalAnyToProto(e)
 		if err != nil {
+			return fmt.Errorf("failed to marshal shim event to any: %w", err)
+		}
+
+		if err := streamer.Send(&types.Envelope{
+			Topic:     runtime.GetTopic(e),
+			Namespace: ns,
+			Event:     evt,
+			Timestamp: timestamppb.Now(),
+		}); err != nil {
 			log.G(ctx).WithError(err).Error("post event")
```

Comment:
If containerd is restarting, original connection has been closed. It seems that the server side is still running. After containerd restarted, there are two connections watching on one channel. If old connection wins, it seems containerd won't get the event. Is it correct? 

Maybe we can fail-fast if connection is closed.

Reply:
Yeah, that's fair. We should exit stream if failed to send event.
Added 1bb75f53a60bcc6ed93e84f3b1e4ed6270beb941

Reply:
I think we still need `Context()` in streamer. When ttrpc connection is closed, we can get notification from context and stop receiving events from channel. It can be done in the follow-up.


![image](https://github.com/containerd/containerd/assets/17510957/274275ec-b221-4b5f-9c2e-0daacb5cae5f)